test:coverage yields errors even though test:unit succeeds

### Problem description
```
yarn test:unit                                    
  1040 passing (4s)
```

```
yarn test:coverage 

  1036 passing (4s)
  4 failing

  1) <CardHeader /> should render CardContent:

      AssertionError: expected 'div' to equal 'withStyles(CardContent)'
      + expected - actual

      -div
      +withStyles(CardContent)
      
      at Context.<anonymous> (src/Card/CardHeader.spec.js:19:12)

  2) <CardMedia /> should have the root class:
     TypeError: ShallowWrapper::dive() can not be called on DOM components
      at ShallowWrapper.<anonymous> (node_modules/enzyme/build/ShallowWrapper.js:1609:19)
      at ShallowWrapper.single (node_modules/enzyme/build/ShallowWrapper.js:1518:25)
      at ShallowWrapper.dive (node_modules/enzyme/build/ShallowWrapper.js:1607:21)
      at shallowWithContext (src/test-utils/createShallow.js:28:22)
      at Context.<anonymous> (src/Card/CardMedia.spec.js:18:21)

  3) <ListItemSecondaryAction /> should render a div:
     TypeError: ShallowWrapper::dive() can not be called on DOM components
      at ShallowWrapper.<anonymous> (node_modules/enzyme/build/ShallowWrapper.js:1609:19)
      at ShallowWrapper.single (node_modules/enzyme/build/ShallowWrapper.js:1518:25)
      at ShallowWrapper.dive (node_modules/enzyme/build/ShallowWrapper.js:1607:21)
      at shallowWithContext (src/test-utils/createShallow.js:28:22)
      at Context.<anonymous> (src/List/ListItemSecondaryAction.spec.js:18:21)

  4) <ListItemSecondaryAction /> should render with the user and root classes:
     TypeError: ShallowWrapper::dive() can not be called on DOM components
      at ShallowWrapper.<anonymous> (node_modules/enzyme/build/ShallowWrapper.js:1609:19)
      at ShallowWrapper.single (node_modules/enzyme/build/ShallowWrapper.js:1518:25)
      at ShallowWrapper.dive (node_modules/enzyme/build/ShallowWrapper.js:1607:21)
      at shallowWithContext (src/test-utils/createShallow.js:28:22)
      at Context.<anonymous> (src/List/ListItemSecondaryAction.spec.js:24:21)

```

### Link to minimal working code that reproduces the issue
Related to flow changes in #7268

Any thoughts on that @oliviertassinari?

--- a/.babelrc
+++ b/.babelrc
@@ -7,6 +7,8 @@
     "coverage": {
       "plugins": [
         [
+          "flow-react-proptypes",
+          "transform-flow-strip-types",
           "istanbul"
         ]
       ]

--- a/package.json
+++ b/package.json
@@ -91,7 +91,7 @@
     "chai": "^4.0.2",
     "cross-env": "^5.0.1",
     "doctrine": "^2.0.0",
-    "enzyme": "^2.8.2",
+    "enzyme": "^2.9.1",
     "eslint": "^3.19.0",
     "eslint-config-airbnb": "^15.0.1",
     "eslint-import-resolver-webpack": "^0.8.1",

--- a/src/AppBar/AppBar.js
+++ b/src/AppBar/AppBar.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -45,7 +45,35 @@ export const styleSheet = createStyleSheet('MuiAppBar', theme => ({
   },
 }));
 
-function AppBar(props) {
+type DefaultProps = {
+  color: 'primary',
+  position: 'fixed',
+};
+
+type Props = DefaultProps & {
+  /**
+   * The content of the component.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The color of the component. It's using the theme palette when that makes sense.
+   */
+  color?: 'inherit' | 'primary' | 'accent' | 'default',
+  /**
+   * The positioning type.
+   */
+  position?: 'static' | 'fixed' | 'absolute',
+};
+
+function AppBar(props: Props) {
   const {
     children,
     classes,
@@ -71,29 +99,6 @@ function AppBar(props) {
     </Paper>
   );
 }
-
-AppBar.propTypes = {
-  /**
-   * The content of the component.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The color of the component. It's using the theme palette when that makes sense.
-   */
-  color: PropTypes.oneOf(['inherit', 'primary', 'accent', 'default']),
-  /**
-   * The positioning type.
-   */
-  position: PropTypes.oneOf(['static', 'fixed', 'absolute']),
-};
 
 AppBar.defaultProps = {
   color: 'primary',

--- a/src/Avatar/Avatar.js
+++ b/src/Avatar/Avatar.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -32,7 +32,64 @@ export const styleSheet = createStyleSheet('MuiAvatar', theme => ({
   },
 }));
 
-function Avatar(props) {
+type DefaultProps = {
+  alt: string,
+  component: string,
+};
+
+type Props = DefaultProps & {
+  /**
+   * Used in combination with `src` or `srcSet` to
+   * provide an alt attribute for the rendered `img` element.
+   */
+  alt?: string,
+  /**
+   * Used to render icon or text elements inside the Avatar.
+   * `src` and `alt` props will not be used and no `img` will
+   * be rendered by default.
+   *
+   * This can be an element, or just a string.
+   */
+  children?: Element<*>,
+  /**
+   * @ignore
+   * The className of the child element.
+   * Used by Chip and ListItemIcon to style the Avatar icon.
+   */
+  childrenClassName?: string,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The component used for the root node.
+   * Either a string to use a DOM element or a component.
+   */
+  component?: string | Function,
+  /**
+   * Properties applied to the `img` element when the component
+   * is used to display an image.
+   */
+  imgProps?: Object,
+  /**
+   * The `sizes` attribute for the `img` element.
+   */
+  sizes?: string,
+  /**
+   * The `src` attribute for the `img` element.
+   */
+  src?: string,
+  /**
+   * The `srcSet` attribute for the `img` element.
+   */
+  srcSet?: string,
+};
+
+function Avatar(props: Props) {
   const {
     alt,
     classes,
@@ -82,58 +139,6 @@ function Avatar(props) {
     </ComponentProp>
   );
 }
-
-Avatar.propTypes = {
-  /**
-   * Used in combination with `src` or `srcSet` to
-   * provide an alt attribute for the rendered `img` element.
-   */
-  alt: PropTypes.string,
-  /**
-   * Used to render icon or text elements inside the Avatar.
-   * `src` and `alt` props will not be used and no `img` will
-   * be rendered by default.
-   *
-   * This can be an element, or just a string.
-   */
-  children: PropTypes.node,
-  /**
-   * @ignore
-   * The className of the child element.
-   * Used by Chip and ListItemIcon to style the Avatar icon.
-   */
-  childrenClassName: PropTypes.string,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The component used for the root node.
-   * Either a string to use a DOM element or a component.
-   */
-  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
-  /**
-   * Properties applied to the `img` element when the component
-   * is used to display an image.
-   */
-  imgProps: PropTypes.object,
-  /**
-   * The `sizes` attribute for the `img` element.
-   */
-  sizes: PropTypes.string,
-  /**
-   * The `src` attribute for the `img` element.
-   */
-  src: PropTypes.string,
-  /**
-   * The `srcSet` attribute for the `img` element.
-   */
-  srcSet: PropTypes.string,
-};
 
 Avatar.defaultProps = {
   alt: '',

--- a/src/Button/Button.js
+++ b/src/Button/Button.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -121,7 +121,77 @@ export const styleSheet = createStyleSheet('MuiButton', theme => ({
   },
 }));
 
-function Button(props) {
+type DefaultProps = {
+  color: 'default',
+  dense: boolean,
+  disabled: boolean,
+  fab: boolean,
+  disableFocusRipple: boolean,
+  raised: boolean,
+  disableRipple: boolean,
+  type: 'button',
+};
+
+type Props = DefaultProps & {
+  /**
+   * The content of the button.
+   */
+  children: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The color of the component. It's using the theme palette when that makes sense.
+   */
+  color?: 'default' | 'inherit' | 'primary' | 'accent' | 'contrast',
+  /**
+   * The component used for the root node.
+   * Either a string to use a DOM element or a component.
+   * The default value is a `button`.
+   */
+  component?: string | Function,
+  /**
+   * Uses a smaller minWidth, ideal for things like card actions.
+   */
+  dense?: boolean,
+  /**
+   * If `true`, the button will be disabled.
+   */
+  disabled?: boolean,
+  /**
+   * If `true`, the  keyboard focus ripple will be disabled.
+   * `ripple` must also be true.
+   */
+  disableFocusRipple?: boolean,
+  /**
+   * If `true`, the ripple effect will be disabled.
+   */
+  disableRipple?: boolean,
+  /**
+   * If `true`, well use floating action button styling.
+   */
+  fab?: boolean,
+  /**
+   * The URL to link to when the button is clicked.
+   * If defined, an `a` element will be used as the root node.
+   */
+  href?: string,
+  /**
+   * If `true`, the button will use raised styling.
+   */
+  raised?: boolean,
+  /**
+   * @ignore
+   */
+  type?: string,
+};
+
+function Button(props: Props) {
   const {
     children,
     classes,
@@ -168,65 +238,6 @@ function Button(props) {
     </ButtonBase>
   );
 }
-
-Button.propTypes = {
-  /**
-   * The content of the button.
-   */
-  children: PropTypes.node.isRequired,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The color of the component. It's using the theme palette when that makes sense.
-   */
-  color: PropTypes.oneOf(['default', 'inherit', 'primary', 'accent', 'contrast']),
-  /**
-   * The component used for the root node.
-   * Either a string to use a DOM element or a component.
-   * The default value is a `button`.
-   */
-  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
-  /**
-   * Uses a smaller minWidth, ideal for things like card actions.
-   */
-  dense: PropTypes.bool,
-  /**
-   * If `true`, the button will be disabled.
-   */
-  disabled: PropTypes.bool,
-  /**
-   * If `true`, the  keyboard focus ripple will be disabled.
-   * `ripple` must also be true.
-   */
-  disableFocusRipple: PropTypes.bool,
-  /**
-   * If `true`, the ripple effect will be disabled.
-   */
-  disableRipple: PropTypes.bool,
-  /**
-   * If `true`, well use floating action button styling.
-   */
-  fab: PropTypes.bool,
-  /**
-   * The URL to link to when the button is clicked.
-   * If defined, an `a` element will be used as the root node.
-   */
-  href: PropTypes.string,
-  /**
-   * If `true`, the button will use raised styling.
-   */
-  raised: PropTypes.bool,
-  /**
-   * @ignore
-   */
-  type: PropTypes.string,
-};
 
 Button.defaultProps = {
   color: 'default',

--- a/src/Card/Card.js
+++ b/src/Card/Card.js
@@ -1,7 +1,6 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -13,28 +12,32 @@ export const styleSheet = createStyleSheet('MuiCard', {
   },
 });
 
-function Card(props) {
+type DefaultProps = {
+  raised: boolean,
+};
+
+type Props = DefaultProps & {
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * If `true`, the card will use raised styling.
+   */
+  raised?: boolean,
+};
+
+function Card(props: Props) {
   const { classes, className, raised, ...other } = props;
 
   return (
     <Paper className={classNames(classes.root, className)} elevation={raised ? 8 : 2} {...other} />
   );
 }
-
-Card.propTypes = {
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * If `true`, the card will use raised styling.
-   */
-  raised: PropTypes.bool,
-};
 
 Card.defaultProps = {
   raised: false,

--- a/src/Card/CardActions.js
+++ b/src/Card/CardActions.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -19,7 +19,30 @@ export const styleSheet = createStyleSheet('MuiCardActions', {
   },
 });
 
-function CardActions(props) {
+type DefaultProps = {
+  disableActionSpacing: boolean,
+};
+
+type Props = DefaultProps & {
+  /**
+   * The content of the component.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * If `true`, the card actions do not have additional margin.
+   */
+  disableActionSpacing?: boolean,
+};
+
+function CardActions(props: Props) {
   const { disableActionSpacing, children, classes, className, ...other } = props;
 
   return (
@@ -30,25 +53,6 @@ function CardActions(props) {
     </div>
   );
 }
-
-CardActions.propTypes = {
-  /**
-   * The content of the component.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * If `true`, the card actions do not have additional margin.
-   */
-  disableActionSpacing: PropTypes.bool,
-};
 
 CardActions.defaultProps = {
   disableActionSpacing: false,

--- a/src/Card/CardContent.js
+++ b/src/Card/CardContent.js
@@ -1,7 +1,6 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -15,21 +14,21 @@ export const styleSheet = createStyleSheet('MuiCardContent', theme => ({
   },
 }));
 
-function CardContent(props) {
+type Props = {
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+};
+
+function CardContent(props: Props) {
   const { classes, className, ...other } = props;
 
   return <div className={classNames(classes.root, className)} {...other} />;
 }
-
-CardContent.propTypes = {
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-};
 
 export default withStyles(styleSheet)(CardContent);

--- a/src/Card/CardHeader.js
+++ b/src/Card/CardHeader.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -22,7 +22,30 @@ export const styleSheet = createStyleSheet('MuiCardHeader', theme => ({
   },
 }));
 
-function CardHeader(props) {
+type Props = {
+  /**
+   * The Avatar  for the Card Header.
+   */
+  avatar?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The content of the component.
+   */
+  subheader?: Element<*>,
+  /**
+   * The content of the Card Title.
+   */
+  title?: Element<*>,
+};
+
+function CardHeader(props: Props) {
   const { avatar, classes, className: classNameProp, subheader, title, ...other } = props;
 
   const className = classNames(classes.root, classNameProp);
@@ -48,28 +71,5 @@ function CardHeader(props) {
     </CardContent>
   );
 }
-
-CardHeader.propTypes = {
-  /**
-   * The Avatar  for the Card Header.
-   */
-  avatar: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The content of the component.
-   */
-  subheader: PropTypes.node,
-  /**
-   * The content of the Card Title.
-   */
-  title: PropTypes.node,
-};
 
 export default withStyles(styleSheet)(CardHeader);

--- a/src/Card/CardMedia.js
+++ b/src/Card/CardMedia.js
@@ -1,7 +1,6 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -12,21 +11,21 @@ export const styleSheet = createStyleSheet('MuiCardMedia', {
   },
 });
 
-function CardMedia(props) {
+type Props = {
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+};
+
+function CardMedia(props: Props) {
   const { classes, className, ...other } = props;
 
   return <div className={classNames(classes.root, className)} {...other} />;
 }
-
-CardMedia.propTypes = {
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-};
 
 export default withStyles(styleSheet)(CardMedia);

--- a/src/Drawer/Drawer.js
+++ b/src/Drawer/Drawer.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React, { Component } from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import Modal from '../internal/Modal';
@@ -74,7 +74,69 @@ export const styleSheet = createStyleSheet('MuiDrawer', theme => ({
   modal: {},
 }));
 
-class Drawer extends Component {
+type DefaultProps = {
+  anchor: 'left',
+  docked: boolean,
+  enterTransitionDuration: number,
+  leaveTransitionDuration: number,
+  open: boolean,
+  elevation: number,
+};
+
+type Props = {
+  /**
+   * Side which will the drawer will appear from.
+   */
+  anchor?: 'left' | 'top' | 'right' | 'bottom',
+  /**
+   * The contents of the drawer.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * If `true`, the drawer will dock itself
+   * and will no longer slide in with an overlay.
+   */
+  docked?: boolean,
+  /**
+   * The elevation of the drawer.
+   */
+  elevation?: number,
+  /**
+   * Customizes duration of enter animation (ms)
+   */
+  enterTransitionDuration?: number,
+  /**
+   * Customizes duration of leave animation (ms)
+   */
+  leaveTransitionDuration?: number,
+  /**
+   * Callback fired when the internal modal requests to be closed.
+   */
+  onRequestClose?: Function,
+  /**
+   * If `true`, the drawer is open.
+   */
+  open?: boolean,
+  /**
+   * Properties applied to the `Slide` element.
+   */
+  SlideProps?: Object,
+};
+
+type State = {
+  firstMount: boolean,
+};
+
+class Drawer extends Component<DefaultProps, Props, State> {
+  props: Props;
   static defaultProps = {
     anchor: 'left',
     docked: false,
@@ -84,7 +146,7 @@ class Drawer extends Component {
     elevation: 16,
   };
 
-  state = {
+  state: State = {
     // Let's assume that the Drawer will always be rendered on user space.
     // We use that state is order to skip the appear transition during the
     // inital mount of the component.
@@ -157,54 +219,6 @@ class Drawer extends Component {
     );
   }
 }
-
-Drawer.propTypes = {
-  /**
-   * Side which will the drawer will appear from.
-   */
-  anchor: PropTypes.oneOf(['left', 'top', 'right', 'bottom']),
-  /**
-   * The contents of the drawer.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * If `true`, the drawer will dock itself
-   * and will no longer slide in with an overlay.
-   */
-  docked: PropTypes.bool,
-  /**
-   * The elevation of the drawer.
-   */
-  elevation: PropTypes.number,
-  /**
-   * Customizes duration of enter animation (ms)
-   */
-  enterTransitionDuration: PropTypes.number,
-  /**
-   * Customizes duration of leave animation (ms)
-   */
-  leaveTransitionDuration: PropTypes.number,
-  /**
-   * Callback fired when the internal modal requests to be closed.
-   */
-  onRequestClose: PropTypes.func,
-  /**
-   * If `true`, the drawer is open.
-   */
-  open: PropTypes.bool,
-  /**
-   * Properties applied to the `Slide` element.
-   */
-  SlideProps: PropTypes.object,
-};
 
 Drawer.contextTypes = {
   styleManager: customPropTypes.muiRequired,

--- a/src/Icon/Icon.js
+++ b/src/Icon/Icon.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import { capitalizeFirstLetter } from '../utils/helpers';
@@ -31,7 +31,30 @@ export const styleSheet = createStyleSheet('MuiIcon', theme => ({
   },
 }));
 
-function Icon(props) {
+type DefaultProps = {
+  color: 'inherit',
+};
+
+type Props = DefaultProps & {
+  /**
+   * The name of the icon font ligature.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The color of the component. It's using the theme palette when that makes sense.
+   */
+  color?: 'inherit' | 'accent' | 'action' | 'contrast' | 'disabled' | 'error' | 'primary',
+};
+
+function Icon(props: Props) {
   const { children, classes, className: classNameProp, color, ...other } = props;
 
   const className = classNames(
@@ -49,33 +72,6 @@ function Icon(props) {
     </span>
   );
 }
-
-Icon.propTypes = {
-  /**
-   * The name of the icon font ligature.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The color of the component. It's using the theme palette when that makes sense.
-   */
-  color: PropTypes.oneOf([
-    'inherit',
-    'accent',
-    'action',
-    'contrast',
-    'disabled',
-    'error',
-    'primary',
-  ]),
-};
 
 Icon.defaultProps = {
   color: 'inherit',

--- a/src/List/List.js
+++ b/src/List/List.js
@@ -73,12 +73,12 @@ type Props = DefaultProps & {
 };
 
 class List extends Component<DefaultProps, Props, void> {
+  props: Props;
   static defaultProps: DefaultProps = {
     component: 'div',
     dense: false,
     disablePadding: false,
   };
-  props: Props;
 
   getChildContext() {
     return {

--- a/src/List/List.js
+++ b/src/List/List.js
@@ -1,4 +1,4 @@
-// @flow weak
+// @flow
 
 import React, { Component } from 'react';
 import type { Element } from 'react';

--- a/src/List/List.js
+++ b/src/List/List.js
@@ -1,6 +1,7 @@
 // @flow weak
 
 import React, { Component } from 'react';
+import type { Element } from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
@@ -27,12 +28,57 @@ export const styleSheet = createStyleSheet('MuiList', theme => ({
   },
 }));
 
-class List extends Component {
-  static defaultProps = {
+type DefaultProps = {
+  component: string,
+  dense: boolean,
+  disablePadding: boolean,
+};
+
+type Props = DefaultProps & {
+  /**
+   * The content of the component.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The component used for the root node.
+   * Either a string to use a DOM element or a component.
+   */
+  component?: string | Function,
+  /**
+   * If `true`, compact vertical padding designed for keyboard and mouse input will be used for
+   * the list and list items. The property is available to descendant components as the
+   * `dense` context.
+   */
+  dense?: boolean,
+  /**
+   * If `true`, vertical padding will be removed from the list.
+   */
+  disablePadding?: boolean,
+  /**
+   * Use that property to pass a ref callback to the root component.
+   */
+  rootRef?: Function,
+  /**
+   * The content of the component, normally `ListItem`.
+   */
+  subheader?: Element<*>,
+};
+
+class List extends Component<DefaultProps, Props, void> {
+  static defaultProps: DefaultProps = {
     component: 'div',
     dense: false,
     disablePadding: false,
   };
+  props: Props;
 
   getChildContext() {
     return {
@@ -70,44 +116,6 @@ class List extends Component {
     );
   }
 }
-
-List.propTypes = {
-  /**
-   * The content of the component.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The component used for the root node.
-   * Either a string to use a DOM element or a component.
-   */
-  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
-  /**
-   * If `true`, compact vertical padding designed for keyboard and mouse input will be used for
-   * the list and list items. The property is available to descendant components as the
-   * `dense` context.
-   */
-  dense: PropTypes.bool,
-  /**
-   * If `true`, vertical padding will be removed from the list.
-   */
-  disablePadding: PropTypes.bool,
-  /**
-   * Use that property to pass a ref callback to the root component.
-   */
-  rootRef: PropTypes.func,
-  /**
-   * The content of the component, normally `ListItem`.
-   */
-  subheader: PropTypes.node,
-};
 
 List.childContextTypes = {
   dense: PropTypes.bool,

--- a/src/List/ListItem.js
+++ b/src/List/ListItem.js
@@ -103,6 +103,7 @@ type Props = DefaultProps & {
 };
 
 class ListItem extends Component<DefaultProps, Props, void> {
+  props: Props;
   static defaultProps: DefaultProps = {
     button: false,
     component: 'div',
@@ -111,7 +112,6 @@ class ListItem extends Component<DefaultProps, Props, void> {
     disableGutters: false,
     divider: false,
   };
-  props: Props;
 
   getChildContext() {
     return {

--- a/src/List/ListItem.js
+++ b/src/List/ListItem.js
@@ -1,6 +1,7 @@
-// @flow weak
+// @flow
 
 import React, { Component } from 'react';
+import type { Element } from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
@@ -52,8 +53,57 @@ export const styleSheet = createStyleSheet('MuiListItem', theme => ({
   },
 }));
 
-class ListItem extends Component {
-  static defaultProps = {
+type DefaultProps = {
+  button: boolean,
+  component: string,
+  dense: boolean,
+  disabled: false,
+  disableGutters: false,
+  divider: false,
+};
+
+type Props = DefaultProps & {
+  /**
+   * If `true`, the ListItem will be a button.
+   */
+  button?: boolean,
+  /**
+   * The content of the component.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The component used for the root node.
+   * Either a string to use a DOM element or a component.
+   */
+  component?: string | Function,
+  /**
+   * If `true`, compact vertical padding designed for keyboard and mouse input will be used.
+   */
+  dense?: boolean,
+  /**
+   * @ignore
+   */
+  disabled?: boolean,
+  /**
+   * If `true`, the left and right padding is removed.
+   */
+  disableGutters?: boolean,
+  /**
+   * If `true`, a 1px light border is added to the bottom of the list item.
+   */
+  divider?: boolean,
+};
+
+class ListItem extends Component<DefaultProps, Props, void> {
+  static defaultProps: DefaultProps = {
     button: false,
     component: 'div',
     dense: false,
@@ -61,6 +111,7 @@ class ListItem extends Component {
     disableGutters: false,
     divider: false,
   };
+  props: Props;
 
   getChildContext() {
     return {
@@ -132,46 +183,6 @@ class ListItem extends Component {
     );
   }
 }
-
-ListItem.propTypes = {
-  /**
-   * If `true`, the ListItem will be a button.
-   */
-  button: PropTypes.bool,
-  /**
-   * The content of the component.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The component used for the root node.
-   * Either a string to use a DOM element or a component.
-   */
-  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
-  /**
-   * If `true`, compact vertical padding designed for keyboard and mouse input will be used.
-   */
-  dense: PropTypes.bool,
-  /**
-   * @ignore
-   */
-  disabled: PropTypes.bool,
-  /**
-   * If `true`, the left and right padding is removed.
-   */
-  disableGutters: PropTypes.bool,
-  /**
-   * If `true`, a 1px light border is added to the bottom of the list item.
-   */
-  divider: PropTypes.bool,
-};
 
 ListItem.contextTypes = {
   dense: PropTypes.bool,

--- a/src/List/ListItemAvatar.js
+++ b/src/List/ListItemAvatar.js
@@ -1,6 +1,7 @@
-// @flow weak
+// @flow
 
 import React from 'react';
+import type { Element } from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import warning from 'warning';
@@ -20,10 +21,25 @@ export const styleSheet = createStyleSheet('MuiListItemAvatar', {
   },
 });
 
+type Props = {
+  /**
+   * The content of the component, normally `Avatar`.
+   */
+  children: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+};
+
 /**
  * It's a simple wrapper to apply the `dense` mode styles to `Avatar`.
  */
-function ListItemAvatar(props, context) {
+function ListItemAvatar(props: Props, context: { dense: boolean }) {
   if (context.dense === undefined) {
     warning(
       false,
@@ -48,21 +64,6 @@ function ListItemAvatar(props, context) {
     ...other,
   });
 }
-
-ListItemAvatar.propTypes = {
-  /**
-   * The content of the component, normally `Avatar`.
-   */
-  children: PropTypes.element.isRequired,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-};
 
 ListItemAvatar.contextTypes = {
   dense: PropTypes.bool,

--- a/src/List/ListItemIcon.js
+++ b/src/List/ListItemIcon.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -15,10 +15,26 @@ export const styleSheet = createStyleSheet('MuiListItemIcon', theme => ({
   },
 }));
 
+type Props = {
+  /**
+   * The content of the component, normally `Icon`, `SvgIcon`,
+   * or a `material-ui-icons` SVG icon component.
+   */
+  children: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+};
+
 /**
  * A simple wrapper to apply `List` styles to an `Icon` or `SvgIcon`.
  */
-function ListItemIcon(props) {
+function ListItemIcon(props: Props) {
   const { children, classes, className: classNameProp, ...other } = props;
 
   return React.cloneElement(children, {
@@ -26,21 +42,5 @@ function ListItemIcon(props) {
     ...other,
   });
 }
-
-ListItemIcon.propTypes = {
-  /**
-   * The content of the component, normally `Icon`, `SvgIcon`,
-   * or a `material-ui-icons` SVG icon component.
-   */
-  children: PropTypes.element.isRequired,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-};
 
 export default withStyles(styleSheet)(ListItemIcon);

--- a/src/List/ListItemIcon.spec.js
+++ b/src/List/ListItemIcon.spec.js
@@ -10,7 +10,7 @@ describe('<ListItemIcon />', () => {
   let classes;
 
   before(() => {
-    shallow = createShallow({ dive: true });
+    shallow = createShallow();
     classes = shallow.context.styleManager.render(styleSheet);
   });
 

--- a/src/List/ListItemSecondaryAction.js
+++ b/src/List/ListItemSecondaryAction.js
@@ -1,7 +1,7 @@
 // @flow weak
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -15,7 +15,22 @@ export const styleSheet = createStyleSheet('MuiListItemSecondaryAction', theme =
   },
 }));
 
-function ListItemSecondaryAction(props) {
+type Props = {
+  /**
+   * The content of the component, normally an `IconButton` or selection control.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+};
+
+function ListItemSecondaryAction(props: Props) {
   const { children, classes, className } = props;
 
   return (
@@ -24,21 +39,6 @@ function ListItemSecondaryAction(props) {
     </div>
   );
 }
-
-ListItemSecondaryAction.propTypes = {
-  /**
-   * The content of the component, normally an `IconButton` or selection control.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-};
 
 ListItemSecondaryAction.muiName = 'ListItemSecondaryAction';
 

--- a/src/Menu/Menu.js
+++ b/src/Menu/Menu.js
@@ -1,8 +1,8 @@
-// @flow weak
+// @flow
 
 import React, { Component } from 'react';
+import type { Element } from 'react';
 import classNames from 'classnames';
-import PropTypes from 'prop-types';
 import { findDOMNode } from 'react-dom';
 import { createStyleSheet } from 'jss-theme-reactor';
 import getScrollbarSize from 'dom-helpers/util/scrollbarSize';
@@ -10,14 +10,80 @@ import Popover from '../internal/Popover';
 import withStyles from '../styles/withStyles';
 import MenuList from './MenuList';
 
+type DefaultProps = {
+  open: boolean,
+  transitionDuration: 'auto',
+};
+
+type Props = DefaultProps & {
+  /**
+   * The DOM element used to set the position of the menu.
+   */
+  anchorEl?: Object,
+  /**
+   * Menu contents, normally `MenuItem`s.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * Properties applied to the `MenuList` element.
+   */
+  MenuListProps?: Object,
+  /**
+   * Callback fired before the Menu enters.
+   */
+  onEnter?: Function,
+  /**
+   * Callback fired when the Menu is entering.
+   */
+  onEntering?: Function,
+  /**
+   * Callback fired when the Menu has entered.
+   */
+  onEntered?: Function, // eslint-disable-line react/sort-prop-types
+  /**
+   * Callback fired before the Menu exits.
+   */
+  onExit?: Function,
+  /**
+   * Callback fired when the Menu is exiting.
+   */
+  onExiting?: Function,
+  /**
+   * Callback fired when the Menu has exited.
+   */
+  onExited?: Function, // eslint-disable-line react/sort-prop-types
+  /**
+   * Callback function fired when the menu is requested to be closed.
+   *
+   * @param {event} event The event that triggered the close request
+   */
+  onRequestClose?: Function,
+  /**
+   * If `true`, the menu is visible.
+   */
+  open?: boolean,
+  /**
+   * The length of the transition in `ms`, or 'auto'
+   */
+  transitionDuration?: number | 'auto',
+};
+
 export const styleSheet = createStyleSheet('MuiMenu', {
   root: {
     maxHeight: 250,
   },
 });
 
-class Menu extends Component {
-  static defaultProps = {
+class Menu extends Component<DefaultProps, Props, void> {
+  static defaultProps: DefaultProps = {
     open: false,
     transitionDuration: 'auto',
   };
@@ -52,7 +118,10 @@ class Menu extends Component {
   handleListKeyDown = (event, key) => {
     if (key === 'tab') {
       event.preventDefault();
-      return this.props.onRequestClose(event);
+      const { onRequestClose } = this.props;
+      if (onRequestClose) {
+        return onRequestClose(event);
+      }
     }
 
     return false;
@@ -118,66 +187,5 @@ class Menu extends Component {
     );
   }
 }
-
-Menu.propTypes = {
-  /**
-   * The DOM element used to set the position of the menu.
-   */
-  anchorEl: PropTypes.object,
-  /**
-   * Menu contents, normally `MenuItem`s.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * Properties applied to the `MenuList` element.
-   */
-  MenuListProps: PropTypes.object,
-  /**
-   * Callback fired before the Menu enters.
-   */
-  onEnter: PropTypes.func,
-  /**
-   * Callback fired when the Menu is entering.
-   */
-  onEntering: PropTypes.func,
-  /**
-   * Callback fired when the Menu has entered.
-   */
-  onEntered: PropTypes.func, // eslint-disable-line react/sort-prop-types
-  /**
-   * Callback fired before the Menu exits.
-   */
-  onExit: PropTypes.func,
-  /**
-   * Callback fired when the Menu is exiting.
-   */
-  onExiting: PropTypes.func,
-  /**
-   * Callback fired when the Menu has exited.
-   */
-  onExited: PropTypes.func, // eslint-disable-line react/sort-prop-types
-  /**
-   * Callback function fired when the menu is requested to be closed.
-   *
-   * @param {event} event The event that triggered the close request
-   */
-  onRequestClose: PropTypes.func,
-  /**
-   * If `true`, the menu is visible.
-   */
-  open: PropTypes.bool,
-  /**
-   * The length of the transition in `ms`, or 'auto'
-   */
-  transitionDuration: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
-};
 
 export default withStyles(styleSheet)(Menu);

--- a/src/Menu/MenuItem.js
+++ b/src/Menu/MenuItem.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import withStyles from '../styles/withStyles';
@@ -28,7 +28,40 @@ export const styleSheet = createStyleSheet('MuiMenuItem', theme => ({
   },
 }));
 
-function MenuItem(props) {
+type DefaultProps = {
+  role: string,
+  selected: boolean,
+};
+
+type Props = DefaultProps & {
+  /**
+   * Menu item contents.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The component used for the root node.
+   * Either a string to use a DOM element or a component.
+   */
+  component?: string | Function,
+  /**
+   * @ignore
+   */
+  role?: string,
+  /**
+   * Use to apply selected styling.
+   */
+  selected?: boolean,
+};
+
+function MenuItem(props: Props) {
   const { classes, className: classNameProp, component, selected, role, ...other } = props;
 
   const className = classNames(
@@ -57,34 +90,6 @@ function MenuItem(props) {
     />
   );
 }
-
-MenuItem.propTypes = {
-  /**
-   * Menu item contents.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The component used for the root node.
-   * Either a string to use a DOM element or a component.
-   */
-  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
-  /**
-   * @ignore
-   */
-  role: PropTypes.string,
-  /**
-   * Use to apply selected styling.
-   */
-  selected: PropTypes.bool,
-};
 
 MenuItem.defaultProps = {
   role: 'menuitem',

--- a/src/Menu/MenuList.js
+++ b/src/Menu/MenuList.js
@@ -1,7 +1,7 @@
-// @flow weak
+// @flow
 
 import React, { Component } from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import { findDOMNode } from 'react-dom';
 import keycode from 'keycode';
 import contains from 'dom-helpers/query/contains';
@@ -9,8 +9,32 @@ import activeElement from 'dom-helpers/activeElement';
 import ownerDocument from 'dom-helpers/ownerDocument';
 import List from '../List';
 
-class MenuList extends Component {
-  state = {
+type Props = {
+  /**
+   * MenuList contents, normally `MenuItem`s.
+   */
+  children?: Element<*>,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * @ignore
+   */
+  onBlur?: Function,
+  /**
+   * @ignore
+   */
+  onKeyDown?: Function,
+};
+
+type State = {
+  currentTabIndex: ?number,
+};
+
+class MenuList extends Component<void, Props, State> {
+  props: Props;
+  state: State = {
     currentTabIndex: undefined,
   };
 
@@ -26,7 +50,7 @@ class MenuList extends Component {
   selectedItem = undefined;
   blurTimer = undefined;
 
-  handleBlur = event => {
+  handleBlur = (event: SyntheticUIEvent) => {
     this.blurTimer = setTimeout(() => {
       if (this.list) {
         const list = findDOMNode(this.list);
@@ -42,7 +66,7 @@ class MenuList extends Component {
     }
   };
 
-  handleKeyDown = event => {
+  handleKeyDown = (event: SyntheticUIEvent) => {
     const list = findDOMNode(this.list);
     const key = keycode(event);
     const currentFocus = activeElement(ownerDocument(list));
@@ -75,7 +99,7 @@ class MenuList extends Component {
     }
   };
 
-  handleItemFocus = event => {
+  handleItemFocus = (event: SyntheticUIEvent) => {
     const list = findDOMNode(this.list);
     if (list) {
       // $FlowFixMe
@@ -123,7 +147,7 @@ class MenuList extends Component {
     return this.setTabIndex(0);
   }
 
-  setTabIndex(index) {
+  setTabIndex(index: number) {
     this.setState({ currentTabIndex: index });
   }
 
@@ -157,24 +181,5 @@ class MenuList extends Component {
     );
   }
 }
-
-MenuList.propTypes = {
-  /**
-   * MenuList contents, normally `MenuItem`s.
-   */
-  children: PropTypes.node,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * @ignore
-   */
-  onBlur: PropTypes.func,
-  /**
-   * @ignore
-   */
-  onKeyDown: PropTypes.func,
-};
 
 export default MenuList;

--- a/src/Paper/Paper.js
+++ b/src/Paper/Paper.js
@@ -1,7 +1,6 @@
-// @flow weak
+// @flow
 
 import React from 'react';
-import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import warning from 'warning';
 import { createStyleSheet } from 'jss-theme-reactor';
@@ -27,7 +26,38 @@ export const styleSheet = createStyleSheet('MuiPaper', theme => {
   };
 });
 
-function Paper(props) {
+type DefaultProps = {
+  component: string,
+  elevation: number,
+  square: boolean,
+};
+
+type Props = DefaultProps & {
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The component used for the root node.
+   * Either a string to use a DOM element or a component.
+   */
+  component?: string | Function,
+  /**
+   * Shadow depth, corresponds to `dp` in the spec.
+   * It's accepting values between 0 and 24 inclusive.
+   */
+  elevation?: number,
+  /**
+   * If `true`, rounded corners are disabled.
+   */
+  square?: boolean,
+};
+
+function Paper(props: Props) {
   const {
     classes,
     className: classNameProp,
@@ -53,31 +83,6 @@ function Paper(props) {
 
   return <ComponentProp className={className} {...other} />;
 }
-
-Paper.propTypes = {
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The component used for the root node.
-   * Either a string to use a DOM element or a component.
-   */
-  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
-  /**
-   * Shadow depth, corresponds to `dp` in the spec.
-   * It's accepting values between 0 and 24 inclusive.
-   */
-  elevation: PropTypes.number,
-  /**
-   * If `true`, rounded corners are disabled.
-   */
-  square: PropTypes.bool,
-};
 
 Paper.defaultProps = {
   component: 'div',

--- a/src/internal/ButtonBase.js
+++ b/src/internal/ButtonBase.js
@@ -1,6 +1,6 @@
 // @flow weak
 import React, { Component } from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import { findDOMNode } from 'react-dom';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
@@ -26,11 +26,69 @@ export const styleSheet = createStyleSheet('MuiButtonBase', {
   },
 });
 
+type DefaultProps = {
+  centerRipple: boolean,
+  focusRipple: boolean,
+  disableRipple: boolean,
+  tabIndex: string,
+  type: string,
+};
+
+type Props = DefaultProps & {
+  centerRipple?: boolean,
+  /**
+   * The content of the component.
+   */
+  children?: Element<*>,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * The component used for the root node.
+   * Either a string to use a DOM element or a component.
+   * The default value is a `button`.
+   */
+  component?: string | Function,
+  /**
+   * If `true`, the base button will be disabled.
+   */
+  disabled?: boolean,
+  /**
+   * If `true`, the ripple effect will be disabled.
+   */
+  disableRipple?: boolean,
+  /**
+   * If `true`, the base button will have a keyboard focus ripple.
+   * `disableRipple` must also be `false`.
+   */
+  focusRipple?: boolean,
+  keyboardFocusedClassName?: string,
+  onBlur?: Function,
+  onClick?: Function,
+  onFocus?: Function,
+  onKeyboardFocus?: Function,
+  onKeyDown?: Function,
+  onKeyUp?: Function,
+  onMouseDown?: Function,
+  onMouseLeave?: Function,
+  onMouseUp?: Function,
+  onTouchEnd?: Function,
+  onTouchStart?: Function,
+  role?: string,
+  tabIndex?: string,
+  type: string,
+};
+
+type State = {
+  keyboardFocused: boolean,
+};
+
 /**
  * @ignore - internal component.
  */
-class ButtonBase extends Component {
-  static defaultProps = {
+class ButtonBase extends Component<DefaultProps, Props, State> {
+  static defaultProps: DefaultProps = {
     centerRipple: false,
     focusRipple: false,
     disableRipple: false,
@@ -38,7 +96,7 @@ class ButtonBase extends Component {
     type: 'button',
   };
 
-  state = {
+  state: State = {
     keyboardFocused: false,
   };
 
@@ -201,6 +259,7 @@ class ButtonBase extends Component {
       classes.root,
       {
         [classes.disabled]: disabled,
+        // $FlowFixMe
         [keyboardFocusedClassName]: keyboardFocusedClassName && this.state.keyboardFocused,
       },
       classNameProp,
@@ -252,52 +311,6 @@ class ButtonBase extends Component {
     );
   }
 }
-
-ButtonBase.propTypes = {
-  centerRipple: PropTypes.bool,
-  /**
-   * The content of the component.
-   */
-  children: PropTypes.node,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * The component used for the root node.
-   * Either a string to use a DOM element or a component.
-   * The default value is a `button`.
-   */
-  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
-  /**
-   * If `true`, the base button will be disabled.
-   */
-  disabled: PropTypes.bool,
-  /**
-   * If `true`, the ripple effect will be disabled.
-   */
-  disableRipple: PropTypes.bool,
-  /**
-   * If `true`, the base button will have a keyboard focus ripple.
-   * `disableRipple` must also be `false`.
-   */
-  focusRipple: PropTypes.bool,
-  keyboardFocusedClassName: PropTypes.string,
-  onBlur: PropTypes.func,
-  onClick: PropTypes.func,
-  onFocus: PropTypes.func,
-  onKeyboardFocus: PropTypes.func,
-  onKeyDown: PropTypes.func,
-  onKeyUp: PropTypes.func,
-  onMouseDown: PropTypes.func,
-  onMouseLeave: PropTypes.func,
-  onMouseUp: PropTypes.func,
-  onTouchEnd: PropTypes.func,
-  onTouchStart: PropTypes.func,
-  role: PropTypes.string,
-  tabIndex: PropTypes.string,
-  type: PropTypes.string,
-};
 
 ButtonBase.contextTypes = {
   styleManager: customPropTypes.muiRequired,

--- a/src/internal/Popover.js
+++ b/src/internal/Popover.js
@@ -1,7 +1,7 @@
 // @flow weak
 
 import React, { Component } from 'react';
-import PropTypes from 'prop-types';
+import type { Element } from 'react';
 import classNames from 'classnames';
 import { createStyleSheet } from 'jss-theme-reactor';
 import contains from 'dom-helpers/query/contains';
@@ -58,8 +58,125 @@ export const styleSheet = createStyleSheet('MuiPopover', {
   },
 });
 
-class Popover extends Component {
-  static defaultProps = {
+type Origin = {
+  horizontal: 'left' | 'center' | 'right' | number,
+  vertical: 'top' | 'center' | 'bottom' | number,
+};
+
+type DefaultProps = {
+  anchorOrigin: Origin,
+  modal: boolean,
+  open: boolean,
+  transformOrigin: Origin,
+  transitionDuration: 'auto',
+  elevation: number,
+};
+
+type Props = DefaultProps & {
+  /**
+   * This is the DOM element that will be used
+   * to set the position of the popover.
+   */
+  anchorEl?: Object,
+  /**
+   * This is the point on the anchor where the popover's
+   * `targetOrigin` will attach to.
+   * Options:
+   * vertical: [top, center, bottom];
+   * horizontal: [left, center, right].
+   */
+  anchorOrigin: Origin,
+  /**
+   * The content of the component.
+   */
+  children?: Element<*>,
+  /**
+   * Useful to extend the style applied to components.
+   */
+  classes: Object,
+  /**
+   * @ignore
+   */
+  className?: string,
+  /**
+   * CSS class or classes applied when the component is entered
+   */
+  elevation?: number,
+  enteredClassName?: string,
+  /**
+   * CSS class or classes applied while the component is entering
+   */
+  enteringClassName?: string,
+  /**
+   * CSS class or classes applied when the component is exited
+   */
+  exitedClassName?: string,
+  /**
+   * CSS class or classes applied while the component is exiting
+   */
+  exitingClassName?: string,
+  /**
+   * @ignore
+   */
+  getContentAnchorEl?: Function,
+  /**
+   * If `true`, the Popover will be rendered as a modal with
+   * scroll locking, focus trapping and a clickaway layer beneath
+   */
+  modal?: boolean,
+  /**
+   * Callback fired before the component is entering
+   */
+  onEnter?: Function,
+  /**
+   * Callback fired when the component is entering
+   */
+  onEntering?: Function,
+  /**
+   * Callback fired when the component has entered
+   */
+  onEntered?: Function, // eslint-disable-line react/sort-prop-types
+  /**
+   * Callback fired before the component is exiting
+   */
+  onExit?: Function,
+  /**
+   * Callback fired when the component is exiting
+   */
+  onExiting?: Function,
+  /**
+   * Callback fired when the component has exited
+   */
+  onExited?: Function, // eslint-disable-line react/sort-prop-types
+  /**
+   * Callback function fired when the popover is requested to be closed.
+   *
+   * @param {event} event The event that triggered the close request
+   */
+  onRequestClose?: Function,
+  /**
+   * If `true`, the popover is visible.
+   */
+  open?: boolean,
+  role?: string,
+  /**
+   * This is the point on the popover which
+   * will attach to the anchor's origin.
+   *
+   * Options:
+   * vertical: [top, center, bottom, x(px)];
+   * horizontal: [left, center, right, x(px)].
+   */
+  transformOrigin: Origin,
+  /**
+   * Set to 'auto' to automatically calculate transition time based on height
+   */
+  transitionDuration: number | 'auto',
+};
+
+class Popover extends Component<DefaultProps, Props, void> {
+  props: Props;
+  static defaultProps: DefaultProps = {
     anchorOrigin: {
       vertical: 'top',
       horizontal: 'left',
@@ -216,10 +333,10 @@ class Popover extends Component {
    * to attach to on the anchor element (or body if none is provided)
    */
   getAnchorOffset(contentAnchorOffset) {
-    const { anchorEl: anchorElProp, anchorOrigin } = this.props;
-
-    const anchorEl = anchorElProp || document.body;
-    const anchorRect = anchorEl.getBoundingClientRect();
+    // $FlowFixMe
+    const { anchorEl, anchorOrigin } = this.props;
+    const anchorElement = anchorEl || document.body;
+    const anchorRect = anchorElement.getBoundingClientRect();
     const anchorVertical = contentAnchorOffset === 0 ? anchorOrigin.vertical : 'center';
 
     return {
@@ -316,108 +433,6 @@ class Popover extends Component {
     );
   }
 }
-
-Popover.propTypes = {
-  /**
-   * This is the DOM element that will be used
-   * to set the position of the popover.
-   */
-  anchorEl: PropTypes.object,
-  /**
-   * This is the point on the anchor where the popover's
-   * `targetOrigin` will attach to.
-   * Options:
-   * vertical: [top, center, bottom];
-   * horizontal: [left, center, right].
-   */
-  anchorOrigin: customPropTypes.origin,
-  /**
-   * The content of the component.
-   */
-  children: PropTypes.node,
-  /**
-   * Useful to extend the style applied to components.
-   */
-  classes: PropTypes.object.isRequired,
-  /**
-   * @ignore
-   */
-  className: PropTypes.string,
-  /**
-   * CSS class or classes applied when the component is entered
-   */
-  elevation: PropTypes.number,
-  enteredClassName: PropTypes.string,
-  /**
-   * CSS class or classes applied while the component is entering
-   */
-  enteringClassName: PropTypes.string,
-  /**
-   * CSS class or classes applied when the component is exited
-   */
-  exitedClassName: PropTypes.string,
-  /**
-   * CSS class or classes applied while the component is exiting
-   */
-  exitingClassName: PropTypes.string,
-  /**
-   * @ignore
-   */
-  getContentAnchorEl: PropTypes.func,
-  /**
-   * If `true`, the Popover will be rendered as a modal with
-   * scroll locking, focus trapping and a clickaway layer beneath
-   */
-  modal: PropTypes.bool,
-  /**
-   * Callback fired before the component is entering
-   */
-  onEnter: PropTypes.func,
-  /**
-   * Callback fired when the component is entering
-   */
-  onEntering: PropTypes.func,
-  /**
-   * Callback fired when the component has entered
-   */
-  onEntered: PropTypes.func, // eslint-disable-line react/sort-prop-types
-  /**
-   * Callback fired before the component is exiting
-   */
-  onExit: PropTypes.func,
-  /**
-   * Callback fired when the component is exiting
-   */
-  onExiting: PropTypes.func,
-  /**
-   * Callback fired when the component has exited
-   */
-  onExited: PropTypes.func, // eslint-disable-line react/sort-prop-types
-  /**
-   * Callback function fired when the popover is requested to be closed.
-   *
-   * @param {event} event The event that triggered the close request
-   */
-  onRequestClose: PropTypes.func,
-  /**
-   * If `true`, the popover is visible.
-   */
-  open: PropTypes.bool,
-  role: PropTypes.string,
-  /**
-   * This is the point on the popover which
-   * will attach to the anchor's origin.
-   *
-   * Options:
-   * vertical: [top, center, bottom, x(px)];
-   * horizontal: [left, center, right, x(px)].
-   */
-  transformOrigin: customPropTypes.origin,
-  /**
-   * Set to 'auto' to automatically calculate transition time based on height
-   */
-  transitionDuration: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
-};
 
 Popover.contextTypes = {
   styleManager: customPropTypes.muiRequired,

